$OpenBSD: patch-src_server_nsca-ng_c,v 1.1 2013/03/14 10:19:38 sthen Exp $

From 979cf50ed690221d2d4ee419cc0f4e67c34179a3 Mon Sep 17 00:00:00 2001
From: Stuart Henderson <stu@spacehopper.org>
Date: Wed, 13 Mar 2013 12:16:40 +0100
Subject: Add a directive to chroot(2) at startup

Allow the user to specify a directory the server should chroot(2) into
on startup.

--- src/server/nsca-ng.c.orig	Tue Mar  5 20:58:00 2013
+++ src/server/nsca-ng.c	Thu Mar 14 10:18:28 2013
@@ -77,7 +77,7 @@ static ev_signal sighup_watcher, sigint_watcher, sigte
 
 static options *get_options(int, char **);
 static void free_options(options *);
-static void switch_user(const char *);
+static void drop_privileges(const char *, const char *);
 static void remove_pidfile(void);
 static void forget_config(void);
 static void signal_cb(EV_P_ ev_signal *, int __attribute__((__unused__)));
@@ -104,8 +104,11 @@ main(int argc, char **argv)
 	cfg = conf_parse(opt->conf_file != NULL ?
 	    opt->conf_file : DEFAULT_CONF_FILE);
 
-	if (cfg_size(cfg, "user") > 0)
-		switch_user(cfg_getstr(cfg, "user"));
+	if (cfg_size(cfg, "user") > 0 || cfg_size(cfg, "chroot") > 0)
+		drop_privileges(cfg_size(cfg, "user") > 0 ?
+		    cfg_getstr(cfg, "user") : NULL,
+		    cfg_size(cfg, "chroot") > 0 ?
+		    cfg_getstr(cfg, "chroot") : NULL);
 
 	if (opt->log_target == -1)
 		opt->log_target = opt->foreground ?
@@ -290,21 +293,25 @@ free_options(options *opt)
 }
 
 static void
-switch_user(const char *user)
+drop_privileges(const char *user, const char *new_root)
 {
 	struct passwd *pw;
 
 	errno = 0;
-	if ((pw = getpwnam(user)) == NULL) {
-		if (errno == 0)
-			die("Cannot find user %s in password database", user);
-		else
-			die("Cannot lookup user %s in password database: %m",
-			    user);
+	if (user != NULL) {
+		if ((pw = getpwnam(user)) == NULL) {
+			if (errno == 0)
+				die("Cannot find user %s", user);
+			else
+				die("Cannot lookup user %s: %m", user);
+		}
+		if (initgroups(user, pw->pw_gid) == -1)
+			die("Cannot set up group list for user %s: %m", user);
 	}
-	if (initgroups(user, pw->pw_gid) == -1
-	    || setgid(pw->pw_gid) == -1
-	    || setuid(pw->pw_uid) == -1)
+	if (new_root != NULL && (chroot(new_root) == -1 || chdir("/") == -1))
+		die("Cannot change root directory to %s: %m", new_root);
+	if (user != NULL
+	    && (setgid(pw->pw_gid) == -1 || setuid(pw->pw_uid) == -1))
 		die("Cannot switch to user %s: %m", user);
 }
 
